models/payment_transaction.py

```diff
@@ -0,0 +1,257 @@
+        _logger.info("sending '/createInvoice' request for link creation:\n%s", pprint.pformat(payload))
+        payment_data = self.provider_id._duitku_make_request('/createInvoice',data=json.dumps(payload),headers=headers)
+
+        self.provider_reference = payment_data.get('reference')
+
+        duitku_tx_values.update({
+            'api_url': payment_data.get('paymentUrl'),
+            # 'payment_url': payment_data['paymentUrl'],
+            'merchantOrderId': processing_values['merchantOrderId'],
+            'expiryPeriod': self.provider_id.duitku_expiry,
+            'reference': payment_data['reference']
+        })
+        return duitku_tx_values
+
+
+#     # --------------------------------------------------
+#     # FORM RELATED METHODS
+#     # --------------------------------------------------
+
+    def _get_tx_from_notification_data(self, provider_code, notification_data):
+        """ Override of `payment` to find the transaction based on APS data.
+
+        :param str provider_code: The code of the provider that handled the transaction.
+        :param dict notification_data: The notification data sent by the provider.
+        :return: The transaction if found.
+        :rtype: recordset of `payment.transaction`
+        :raise ValidationError: If inconsistent data are received.
+        :raise ValidationError: If the data match no transaction.
+        """
+        tx = super()._get_tx_from_notification_data(provider_code, notification_data)
+        if provider_code != 'duitku' or len(tx) == 1:
+            return tx
+
+        reference = notification_data.get('merchantOrderId')
+        if not reference:
+            raise ValidationError(
+                "Duitku: " + _("Received data with missing reference %(ref)s.", ref=reference)
+            )
+
+        tx = self.search([('reference', '=', reference), ('provider_code', '=', 'duitku')])
+        if not tx or len(tx) > 1:
+            error_msg = 'Duitku: received data for reference %s' % (reference)
+            if not tx:
+                error_msg += '; no order found'
+            else:
+                error_msg += '; multiple order found'
+            raise ValidationError("Duitku: " + error_msg)
+
+        return tx
+
+    def _process_notification_data(self, notification_data):
+        """ Override of payment to process the transaction based on Duitku data.
+
+        Note: self.ensure_one()
+
+        :param dict notification_data: The notification data sent by the provider
+        :return: None
+        """
+        super()._process_notification_data(notification_data)
+        if self.provider_code != 'duitku':
+            return
+
+        self.duitku_reference = self.provider_reference= notification_data.get('reference')
+        self.duitku_order_id= notification_data.get('merchantOrderId')
+
+        payment_status = notification_data.get('resultCode')
+
+        if payment_status == '00':
```

Comment:
Hi @wm-cortex  , Please add signature validation and check transaction validation before processing here.